src/step-function.ts

```diff
@@ -297,40 +278,52 @@ export namespace $SFN {
    * ```
    * @param paths
    */
-  // @ts-ignore
-  export function parallel<Paths extends readonly (() => any)[]>(
-    ...paths: Paths
-  ): {
-    [i in keyof Paths]: i extends `${number}`
-      ? ReturnType<Extract<Paths[i], () => any>>
-      : Paths[i];
-  };
-
-  export function parallel(call: CallExpr, context: CallContext): ParallelTask {
-    if (context.kind !== ASL.ContextName) {
-      throw new Error(`$SFN.${kind} is only supported by ${ASL.ContextName}`);
-    }
-    const paths = call.getArgument("paths")?.expr;
-    if (paths === undefined) {
-      throw new Error("missing required argument 'paths'");
-    }
-    if (paths.kind !== "ArrayLiteralExpr") {
-      throw new Error(`invalid arguments to $SFN.parallel`);
+  export const parallel = makeIntegration<
+    <Paths extends readonly (() => any)[]>(
+      ...paths: Paths
+    ) => {
+      [i in keyof Paths]: i extends `${number}`
+        ? ReturnType<Extract<Paths[i], () => any>>
+        : Paths[i];
     }
-    ensureItemOf(
-      paths.items,
-      isFunctionExpr,
-      `each parallel path must be an inline FunctionExpr`
-    );
+  >({
+    ...sfnIntegrationBase("parallel"),
+    asl(call, context) {
+      const paths = call.getArgument("paths")?.expr;
+      if (paths === undefined) {
+        throw new Error("missing required argument 'paths'");
+      }
+      if (paths.kind !== "ArrayLiteralExpr") {
+        throw new Error(`invalid arguments to $SFN.parallel`);
+      }
+      ensureItemOf(
+        paths.items,
+        isFunctionExpr,
+        `each parallel path must be an inline FunctionExpr`
+      );
 
-    return {
-      Type: "Parallel",
-      Branches: paths.items.map((func) => ({
-        StartAt: context.getStateName(func.body.step()!),
-        States: context.execute(func.body),
-      })),
-    };
-  }
+      return {
+        Type: "Parallel",
+        Branches: paths.items.map((func) => ({
+          StartAt: context.getStateName(func.body.step()!),
+          States: context.execute(func.body),
+        })),
+      };
+    },
+  });
+}
+
+function sfnIntegrationBase(
+  methodName: string
+): Pick<IIntegration, "kind" | "unhandledContext"> {
+  return {
+    kind: `$SFN.${methodName}`,
+    unhandledContext(kind, context) {
+      throw new Error(
+        `${kind} is only allowed within a '${VTL.ContextName}' context, but was called within a '${context.kind}' context.`
+      );
+    },
+  };
 }
```

Comment:
This is a poor-mans abstract class. Why not just use a class? Let's try and keep the coding patterns simple to understand and read. Use conventional patterns.

```ts
export abstract class BaseStepFunctionIntegration extends BaseIntegration {
  constructor(readonly methodName: string) {
    super(`$SFN.${methodName}`);
  }

  public abstract asl(call: CallExpr, context: ASL): State;
}
```

Reply:
We'll see, I am trying to avoid working with classes to keep the closure serialization simpler, also keeps the context of `this` simpler. I generally prefer simple objects with interfaces over classes. I'll try with the wrapper function detailed here https://github.com/sam-goodwin/functionless/pull/108#discussion_r859514099. This should remove the need to spread the resulting object.